### ISSUE TITLE
Split download in parallel to speed up task for limited speed servers

### ISSUE THREAD

**Issue by nosmirck** [781flyingdutchman/background_downloader#111]
Hello, first of all thank you for this amazing work, I really appreciate your effort on this package. If you have any way to accept contributions (like donations, buy a coffee, Patreon, etc) please let me know!

Now, on the subject.

I'm working on a file downloader for Archive.org. it's well known that they cap the download speeds, however, using something like "Free Download Manager (FDM)" you can speed up the download, which is possible with FDM because it divides the download in chunks that downloads in parallel effectively speeding up the download by 2X, 3X, 4X etc (depends on how many chunks it does).

My question is, does this package support downloads by chunks in parallel? If so, how?

And if it's possible, can we set different URLs for different chunks? For instance, in archive.org they usually place the same file in at least 2 different servers, so you can have 4 to 8 chunks at the same time (2 to 4 on each server) since the servers also have a limitation of 2 or 4 connections per client. So, if the download has a limit of 100Kb/s you could speed up the download to almost 1Mb/s using 8 chunks.

If this feature is not available, that would be then a request, to be able to set multiple URLs to a task and a number of chunks to distribute among the URLs (like, let's say I have 3 servers with the same file that allow 4 connections each, I can set the chunks to anything from 1, which will use 1 server, up to 12 chunks which will get up to the full speed possible).

Thanks in advance!

**Comment by 781flyingdutchman** [comment]
Hi, thanks!  Interesting request - I'm working on a few other things at the moment, but I'll leave this open and get back to it. Two questions: 1) what platform(s) (iOS, Android, Desktop) are you looking to do this for, and 2) if I implement this, can you commit to testing this feature on all platforms (as I can't)?

Thanks!

**Comment by nosmirck** [comment]
@781flyingdutchman hello and thanks for the quick response!

To answer your questions:

1) this app is targeted for Android and Linux at the moment, but will eventually be available for iOS, Windows and MacOS. Web is not in the plans.
2) absolutely, I have several devices (android with versions 9, 10 and 12, between phones, tablets and a couple of retro gaming devices), couple of iOS devices (X and 13), Linux machines (an old laptop with Lubuntu and a Steam Deck which is my primary target for Linux), two windows laptops (with 10 and 11) and finally an old Intel Mac (I plan on getting an M2 machine soon).

I'll take a look at the code and see if I can help with it too.

**Comment by nosmirck** [comment]
Sorry for the late reply! I was on vacations!

here you can find a sample json metadata for some NASA videos on Archive.org.

https://archive.org/metadata/jsc2022m000172_Earth_in_4K_Expedition_65_Edition

You can see that all the files in here are mirrored on 2 servers
```
  "d1": "ia601507.us.archive.org",
  "d2": "ia801507.us.archive.org",
```

So, for instance, this file `jsc2022m000172_Earth_in_4K_Expedition_65_Edition.mp4`

can be downloaded with these two urls:

```
https://ia801507.us.archive.org/8/items/jsc2022m000172_Earth_in_4K_Expedition_65_Edition/jsc2022m000172_Earth_in_4K_Expedition_65_Edition.mp4

https://ia601507.us.archive.org/8/items/jsc2022m000172_Earth_in_4K_Expedition_65_Edition/jsc2022m000172_Earth_in_4K_Expedition_65_Edition.mp4
```

These files are very big and archive.org servers have throttle speeds, they allow for portioned downloads similar to what Free Download Manager (FDM) does, but limiting the number of connections depending on demand (anywhere from 1 to 8 simultaneous connections) but adding a mirror will guarantee that at least 2 connections are stablished, effectively duplicating the download speed (assuming of course your bandwidth is greater than the limitations from the server).

**Comment by 781flyingdutchman** [comment]
Hi @nosmirck , wondering if you can help test the initial implementation (Desktop only). This introduces a new task type, `ParallelDownloadTask`, which you define for instance as follows:
```
task = ParallelDownloadTask(
          url: [url1, url2], filename: 'file.dat', chunks: 2);
     
```

The above would create a task with 4 chunks: 2 urls and 2 chunks per url (you can also just pass one url, not in a list, and you'll get 2 chunks for the same url).  The rest is basically the same: call `enqueue` or `download` and monitor or wait for the result. Remember if you want progress updates you need to set the task's `updates` field to `Updates.statusAndProgress`.

You can set `retries` on the task (which translates into retries for every 'chunk' download task), and you can cancel (which cancels every chunk download). You cannot currently pause/resume. I'd like your thoughts on how big a deal this is, as it's somewhat challenging to implement.

How it works is:
* We make a HEAD request to get content length
* We chunk the download into chunks of equal size
* We create a regular DownloadTask for each chunk, with the headers containing a 'Range: bytes=from-to' entry, and the group set to 'chunk'
* We 'spawn' these child tasks, linked to the parent task
* We intercept status and progress for task in the 'chunk' group and pass this to the parent, where we aggregate status/progress across all chunk tasks and act accordingly (eg fail all on failure of one, stitch together when all are complete, etc).

Please give it a try and let me know if you run into any issues (there is rather exhaustive logging, so please include that if you run into issues). I'd like to make sure that this works well for Desktop before translating this to a similar mechanism in Android and iOS.

Thanks for your help and patience.

**Comment by 781flyingdutchman** [comment]
Hi @nosmirck have you had an opportunity to test? It's proven to be quite complicated to make this work across all platforms, so I am beginning to wonder whether it's worth doing it.

**Comment by 781flyingdutchman** [comment]
Hi @nosmirck can you please take a look at the latest, found on branch `parallel` which you can get by changing your pubspec.yaml to:
```
background_downloader:
    git:
      url: https://github.com/781flyingdutchman/background_downloader.git
      ref: parallel
```

At the start of this bug you confirmed you'd be able to test on Android, iOS and Desktop, and parallel downloads should now work on all platforms, so please test and let me know if/where you run into issues.  Your help in testing is critical, and was kinda a prerequisite for me investing the time and energy to develop this feature, so your help is also much appreciated.

**Comment by 781flyingdutchman** [comment]
Published in 7.10.0
